### ISSUE TITLE
Potential bug in Porter stemmer

### ISSUE THREAD

**Issue by AmenRa** [nltk/nltk#2623]
Hi, I accidentally noticed that at line [504](https://github.com/nltk/nltk/blob/385fb3f22031ebfa87b5c5571f32008f967264fa/nltk/stem/porter.py#L504) of porter.py there is a potential bug:
```python
rules.append(
    ("logi", "log", lambda stem: self._has_positive_measure(word[:-3]))
)
```
The `lambda` function's body refers to `word` and not to `stem`.
Is it correct or should it refer to `stem` instead?

**Comment by stevenbird** [comment]
Yes, that's a bug! Well spotted.

**Comment by tomaarsen** [comment]
@AmenRa @stevenbird 

This appears to *not* be a bug. Below is a short explanation as to why, as well as some tests showing that the current behaviour is desirable, and the behaviour after applying a "fix" is not.

---

### Explanation
The comment above the line in question clarifies the reasoning somewhat:
https://github.com/nltk/nltk/blob/860c494a97c73ee3881a6c57e6468ad888778ad5/nltk/stem/porter.py#L500-L505

The normal rules that are appended are in the form:
```python
            ("ational", "ate", self._has_positive_measure),
```
For `suffix`, `replacement` and `condition` respectively.
In this case any given word is tested whether it has the suffix `"ational"`. If it does, this suffix is removed, producing a `stem`. This stem, e.g. `"r"` (for `"rational"`) is then passed to the condition: `self._has_positive_measure`, which states whether `stem + replacement` should be returned, or not. 

However, as the comment says, the condition for words ending in `"logi"` (e.g. `"theologi"`, `"archaeologi"`, `"geologi"`) must check the condition with the `"l"` included in the stem, i.e. test the condition with `"theol"` instead of `"theo"`, and `"geol"` instead of `"geo"`. 
That is why
```python
# Note: "theologi"[:-3] == "theol"
("logi", "log", lambda stem: self._has_positive_measure(word[:-3]))
```
is used, rather than 
```python
("logi", "log", self._has_positive_measure)
```

---

### Tests:
I've ran the following program first on the current `develop` branch. (Python 3.7+ because f-string formatting)
```python
from nltk.stem import PorterStemmer

ps = PorterStemmer()
def test(word):
    print(f"{word:<12} -> {ps.stem(word)}")

test("philologi")
test("archaeologi")
test("theologi")
test("geologi")
```
Which produces:
```python
philologi    -> philolog
archaeologi  -> archaeolog
theologi     -> theolog
geologi      -> geolog
```
This is correct in all cases.

After applying the expected "fix" by replacing the line in question with:
```python
            rules.append(
                ("logi", "log", lambda stem: self._has_positive_measure(stem[:-3]))
            )
```
The output becomes:
```python
philologi    -> philologi
archaeologi  -> archaeolog
theologi     -> theologi
geologi      -> geologi
```
Which is incorrect in 3 out of the 4 cases.

Replacing `lambda stem: self._has_positive_measure(wrd[:-3])` with just `self._has_positive_measure` is also wrong 2 out of the 4 times. Exactly as the comment says, `"geologi"` and `"theologi"` are not handled like `"archaeologi"`, and are wrongly stemmed.

---

In short, this is intentional. No bug here.
